### ISSUE TITLE
Test failures are silenced when Rscript and try(test_that()) are combined

### ISSUE THREAD

**Issue by fritzo** [nimble-dev/nimble#395]
## Overview

Many of our tests on travis-ci are (and always have been) silenced. To fix this, we can set `runViaTestthat` in [`run_tests.R`](https://github.com/nimble-dev/nimble/blob/devel/run_tests.R). However because some of our tests are currently failing, this will require either fixing those tests or `skip()`ping those tests as `KNOWN_FAILURES`.

## Detailed diagnosis

When `test_that()` is wrapped in a `try()`, then test failures will be reported by `test_package('nimble', 'myFile')` but are silenced by `Rscript inst/tests/test-myFile.R`. Specifically, `Rscript ...` will print the errors but then continue to subsequent tests and exit with successful return code 0. This is problematic because our current travis tests are (and always have been) run via `Rscript` rather than `testthat::test_package()`.

It is probable that developers adopted the `try(test_that(-))` idiom to avoid stopping at the first failure when running under `Rscript` (or when sourcing an entire file in an IDE). While this idiom is helpful for manual inspection, it has led to silenced errors when tests are run automatically.

As of 25f24b1, this repo contains 177 occurrences of the `try(test_that())` idiom with discarded return value, spread over 11 files.
```sh
$ find . -name '*.R' \
  | xargs grep -P '^\s*try\([\s\n]*test_that' -hc \
  | awk '{s+=$1} END {print s}'
177

$ find . -name '*.R' \
  | xargs grep -P '^\s*try\([\s\n]*test_that' -c \
  | grep -v ':0$' \
  | sed 's/\(.*\):\(.*\)/\2\t\1/g' \
  | sort -nr
41	./inst/tests/test-distributions.R
38	./inst/tests/test-trunc.R
36	./inst/tests/test-getDistributionInfo.R
16	./inst/tests/test-dsl_dists.R
13	./inst/tests/test_utils.R
11	./inst/tests/test-user.R
8	./inst/tests/test-models.R
6	./inst/tests/test-nimbleFunctionInterfaces.R
5	./inst/tests/test-mcem.R
2	./inst/tests/test-mcmc.R
1	./inst/tests/test-modelValuesInterface.R
```

## Possible fixes

1. Run all travis tests through `testthat::test_package`. This can be done simply by setting setting `runViaTestthat = TRUE` in [`run_tests.R`](https://github.com/nimble-dev/nimble/blob/devel/run_tests.R).

2. We may be able to work around `testthat`'s dangerous behavior by setting a global test context in [`test_utils.R`](https://github.com/nimble-dev/nimble/blob/devel/packages/nimble/inst/tests/test_utils.R). If this is possible, it has the advantage that developers could avoid the `try(test_that(-))` idiom while still being able to collect multiple errors (rather than stopping at the first error).

Both of these solutions require eventual cleanup of our currently silently failing tests (e.g. `test-sizes.R` of #363). An immediate patch and workflow is to `skip()` those tests as `KNOWN_FAILURES`, e.g.
```r
# In test_utils.R:
KNOWN_FAILURE <- function() {
    if(length(Sys..getenv('RUN_KNOWN_FAILURES')) == 0) skip('KNOWN_FAILURE')
}

# In test-sizes.R or another test file:
test('Some test that is known to fail', {
    KNOWN_FAILURE()
    # Rest of test code follows...
})
```
Then our workflow for fixing known failures will be to run
```sh
RUN_KNOWN_FAILURES=true ./run_tests.R
```

**Comment by perrydv** [comment]
Wow, thanks @fritzo!  It seems the most practical way to get out 0.6-5 will be to use `runViaTestthat` on everything and as you suggest `skip` the `KNOWN_FAILURE` cases.  This would be followed by thorough clean-up after release.  Would you agree?  I'm making progress on the nimbleList problem, which I guess illustrates one kind of testing hazard.

**Comment by fritzo** [comment]
@perrydv I think 0.6-5 can safely release before this fix. I believe this issue is not blocking our CRAN submission; it is merely internal testing health. There will probably be a bit of work wiring `skip()` into our existing test helpers (in `test_utils.R`), and I wouldn't want to block our CRAN submission by my slow progress on this issue.

**Comment by perrydv** [comment]
That's good news.  But are we confident there aren't silenced test failures we should be addressing?

**Comment by fritzo** [comment]
That's a great question and I don't know the answer. But I believe this issue is nearly 2 years old (since June 10, 2015), so I'm inclined to consider it out-of-scope for 0.6-5. What do you think @paciorek ?

**Comment by paciorek** [comment]
Yes, I think we should move forward with 0.6-5 (but see another email I
will send) and clean up our testing in coming weeks.

**Comment by perrydv** [comment]
Now that I have dived more deeply into what is going on, here are some comments for posterity:

* In the large majority (perhaps nearly all?) of cases, the "silent failures" were known to the development team and having them in the system was simply a non-smooth testing system.  Cleaning these up will help us be sure to see new failures and work more efficiently.  

* The `try(test_that(...))` idiom can be used successfully without silencing errors if the various outcomes are managed carefully.  Moving towards always running via `test_package` and avoiding the `try(test_that(...))` idiom will be a good move.

* An approach (`testIfNotKnownFailure(knownFailure, expectation)`) for managing known errors when they may appear in long test functions with lots of `expect_*()` calls is [here](https://github.com/nimble-dev/nimble/blob/23efa0e30eb24e95d0bb856c1272abe895946d46/packages/nimble/inst/tests/test_utils.R#L269), with use illustrated [here](https://github.com/nimble-dev/nimble/blob/23efa0e30eb24e95d0bb856c1272abe895946d46/packages/nimble/inst/tests/test_utils.R#L311).

**Comment by fritzo** [comment]
This was mostly fixed by #402 and we now run via `test_package` on travis.
